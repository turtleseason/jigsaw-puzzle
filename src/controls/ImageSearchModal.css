.search-results {
    column-count: 1;
    column-gap: .75rem;
}

@media (min-width: 992px) {
    .search-results {
        column-count: 2;
    }
}

.result-container {
    display: inline-block;
    width: 100%;
}

.result {
    border: 3px solid transparent;
    width: 100%;
    height: 100%;
}

.result-selected {
    border-color: var(--primary, #007bff);
}

.result img {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
}

.result .placeholder {
    position: relative;
    width: 100%;
    height: 100%;
    background: #adb5bd;
}

.result .badge {
    position: absolute;
    bottom: .5rem;
    left: .5rem;
}

.search-error h6 {
    font-size: 1.25rem;
}

.form-control.is-invalid {
    background-image: none !important;
}