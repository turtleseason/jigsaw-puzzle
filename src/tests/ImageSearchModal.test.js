import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { render, screen, waitFor, waitForElementToBeRemoved, getByRole } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import ImageSearchModal from '../controls/ImageSearchModal.js';
import { testSearchResults } from './testData/testSearchResults.js';


const proxyUrl = process.env.REACT_APP_PROXY_URL;

const searchURL = `${proxyUrl}/search/:query/page/:pageNumber`;
const downloadURL = `${proxyUrl}/download/:photoId/`;

const server = setupServer(
    rest.get(searchURL, (req, res, ctx) =>
        res(ctx.status(200),
            ctx.json({
                status: 200,
                type: 'success',
                response: {
                    results: testSearchResults[req.params.pageNumber - 1],
                    total: 17,
                    total_pages: 2,
                },
            })
        )
    ),
    rest.get(downloadURL, (req, res, ctx) =>
        res(ctx.status(200), ctx.json({ type: 'success' }))
    )
);

beforeAll(() => server.listen({ onUnhandledRequest: 'error' }));
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test('renders without exploding', async () => {
    render(<ImageSearchModal isOpen={true} setImage={() => { }} generateDefaultDimensions={() => { }} toggleModal={() => { }} />);
});

test('happy path works', async () => {
    const mockSetImage = jest.fn();
    const mockDimensions = jest.fn(() => { return { rows: 4, cols: 2 }; });

    render(
        <ImageSearchModal
            isOpen={true}
            setImage={mockSetImage}
            generateDefaultDimensions={mockDimensions}
            toggleModal={() => { }}
        />
    );

    userEvent.type(screen.getByRole('textbox', { name: /Search by keyword/i }), 'cat');
    userEvent.click(screen.getByRole('button', { name: /Search/i }));

    const results = await screen.findAllByTestId('result');

    userEvent.click(results[0]);
    userEvent.click(results[1]);

    expect(screen.getByRole('dialog')).toMatchSnapshot('snapshot with search results');

    userEvent.click(screen.getByRole('button', { name: /Use selected image/i }));

    const expected = testSearchResults[0][1];
    expect(mockSetImage).toHaveBeenCalledTimes(1);
    expect(mockSetImage).toHaveBeenCalledWith(
        expect.objectContaining({
            name: expect.stringContaining('Unsplash Photo'),
            url: expected.urls.regular,
            author: expected.user.name,
            defaultRows: 4,
            defaultCols: 2,
            source: expect.objectContaining({
                sourceName: expect.stringMatching(/Unsplash/i),
                nameIsTitle: false,
                isPhoto: true,
                sourceUrl: expected.links.html,
            }),
        })
    );
});

test('happy path works (with page navigation)', async () => {
    const mockSetImage = jest.fn();
    const mockDimensions = jest.fn(() => { return { rows: 4, cols: 2 }; });
    const mockScroll = jest.fn();

    // Stub DOM method that doesn't exist in the testing environment
    window.HTMLElement.prototype.scrollIntoView = mockScroll;

    render(
        <ImageSearchModal
            isOpen={true}
            setImage={mockSetImage}
            generateDefaultDimensions={mockDimensions}
            toggleModal={() => { }}
        />
    );

    userEvent.type(screen.getByRole('textbox', { name: /Search by keyword/i }), 'cat');
    userEvent.click(screen.getByRole('button', { name: /Search/i }));

    const firstPageResults = await screen.findAllByTestId('result');

    userEvent.click(await screen.findByRole('button', { name: '2' }));

    await waitForElementToBeRemoved(firstPageResults[0]);

    const results = await screen.findAllByTestId('result');
    userEvent.click(results[2]);

    expect(screen.getByRole('dialog')).toMatchSnapshot('snapshot on page 2 of search results');

    userEvent.click(screen.getByRole('button', { name: /Use selected image/i }));

    const expected = testSearchResults[1][2];
    expect(mockSetImage).toHaveBeenCalledTimes(1);
    expect(mockSetImage).toHaveBeenCalledWith(
        expect.objectContaining({
            name: expect.stringContaining('Unsplash Photo'),
            url: expected.urls.regular,
            author: expected.user.name,
            defaultRows: 4,
            defaultCols: 2,
            source: expect.objectContaining({
                sourceName: expect.stringMatching(/Unsplash/i),
                nameIsTitle: false,
                isPhoto: true,
                sourceUrl: expected.links.html,
            }),
        })
    );
    expect(mockScroll).toHaveBeenCalledTimes(1);
});

test('renders correctly with no results', async () => {
    server.use(rest.get(searchURL, (req, res, ctx) =>
        res(ctx.status(200), ctx.json({
            status: 200,
            type: 'success',
            response: {
                results: [],
                total: 0,
                total_pages: 0,
            },
        }))
    ));

    render(<ImageSearchModal isOpen={true} setImage={() => { }} generateDefaultDimensions={() => { }} toggleModal={() => { }} />);

    userEvent.type(screen.getByRole('textbox', { name: /Search by keyword/i }), 'cat');
    userEvent.click(screen.getByRole('button', { name: /Search/i }));

    expect(await screen.findByText(/No results found/)).toBeVisible();
    expect(screen.getByRole('dialog')).toMatchSnapshot();
});

test('search fails if empty/whitespace', async () => {
    server.use(rest.get(searchURL, () => { throw Error('Unexpected network request'); }));

    render(<ImageSearchModal isOpen={true} setImage={() => { }} generateDefaultDimensions={() => { }} toggleModal={() => { }} />);

    const input = screen.getByRole('textbox', { name: /Search by keyword/i });

    // Would be good to rewrite the form so that the validity state is more accessible instead of checking for a class
    //
    // In theory, another way to test would be to check the presence of the "Please enter a search term" error text,
    // but that doesn't work (seems to be because the CSS that hides/shows it isn't loaded in the testing environment)
    expect(input).not.toHaveClass('is-invalid');

    userEvent.click(screen.getByRole('button', { name: /Search/i }));

    expect(input).toHaveValue('');
    expect(input).toHaveClass('is-invalid');

    userEvent.type(screen.getByRole('textbox', { name: /Search by keyword/i }), '    ');
    userEvent.click(screen.getByRole('button', { name: /Search/i }));

    expect(input).toHaveValue('    ');
    expect(input).toHaveClass('is-invalid');
});

test('search escapes special characters', async () => {
    const query = 'q:u/e&r?=y';
    const escapedQuery = encodeURIComponent(query);

    let searchedQuery = '';

    server.use(rest.get(searchURL, (req, res, ctx) => {
        searchedQuery = req.params.query;
        return res(ctx.status(200), ctx.json({
            status: 200,
            type: 'success',
            response: {
                results: testSearchResults[req.params.pageNumber - 1],
                total: 17,
                total_pages: 2,
            }
        }));
    }));

    render(<ImageSearchModal isOpen={true} setImage={() => { }} generateDefaultDimensions={() => { }} toggleModal={() => { }} />);

    userEvent.type(screen.getByRole('textbox', { name: /Search by keyword/i }), query);
    userEvent.click(screen.getByRole('button', { name: /Search/i }));

    await waitFor(() => expect(searchedQuery).toBe(escapedQuery));
});

test('renders error message correctly (proxy error)', async () => {
    server.use(rest.get(searchURL, (req, res, ctx) => res(ctx.status(400))));

    render(<ImageSearchModal isOpen={true} setImage={() => { }} generateDefaultDimensions={() => { }} toggleModal={() => { }} />);

    userEvent.type(screen.getByRole('textbox', { name: /Search by keyword/i }), 'cat');
    userEvent.click(screen.getByRole('button', { name: /Search/i }));

    const errorMessage = await screen.findByRole('alert');
    expect(errorMessage).toBeVisible();
    expect(errorMessage).toMatchSnapshot();
});

test('renders error message correctly (upstream error)', async () => {
    server.use(
        rest.get(searchURL, (req, res, ctx) =>
            res(ctx.status(200), ctx.json({
                status: 400,
                type: 'failure',
                errors: ['Pretend something bad happened'],
            }))
        ),
    );

    render(<ImageSearchModal isOpen={true} setImage={() => { }} generateDefaultDimensions={() => { }} toggleModal={() => { }} />);

    userEvent.type(screen.getByRole('textbox', { name: /Search by keyword/i }), 'cat');
    userEvent.click(screen.getByRole('button', { name: /Search/i }));

    const errorMessage = await screen.findByRole('alert');
    expect(errorMessage).toBeVisible();
    expect(errorMessage).toMatchSnapshot();
});

test('renders rate limit error message correctly (proxy)', async () => {
    server.use(rest.get(searchURL, (req, res, ctx) => res(ctx.status(429))));

    render(<ImageSearchModal isOpen={true} setImage={() => { }} generateDefaultDimensions={() => { }} toggleModal={() => { }} />);

    userEvent.type(screen.getByRole('textbox', { name: /Search by keyword/i }), 'cat');
    userEvent.click(screen.getByRole('button', { name: /Search/i }));

    const errorMessage = await screen.findByRole('alert');
    expect(errorMessage).toBeVisible();
    expect(getByRole(errorMessage, 'heading')).toHaveTextContent(/Rate limit exceeded/i);
    expect(errorMessage).toMatchSnapshot();
});

test('renders rate limit error message correctly (upstream)', async () => {
    server.use(
        rest.get(searchURL, (req, res, ctx) =>
            res(ctx.status(200), ctx.json({
                status: 429,
                type: 'failure',
                errors: ['Rate limit exceeded'],
            }))
        ),
    );

    render(<ImageSearchModal isOpen={true} setImage={() => { }} generateDefaultDimensions={() => { }} toggleModal={() => { }} />);

    userEvent.type(screen.getByRole('textbox', { name: /Search by keyword/i }), 'cat');
    userEvent.click(screen.getByRole('button', { name: /Search/i }));

    const errorMessage = await screen.findByRole('alert');
    expect(errorMessage).toBeVisible();
    expect(getByRole(errorMessage, 'heading')).toHaveTextContent(/Rate limit exceeded/i);
    expect(errorMessage).toMatchSnapshot();
});