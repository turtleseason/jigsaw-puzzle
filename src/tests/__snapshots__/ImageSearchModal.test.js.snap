// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`happy path works (with page navigation): snapshot on page 2 of search results 1`] = `
<div
  class="modal fade"
  role="dialog"
  style="display: block;"
  tabindex="-1"
>
  <div
    class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
    role="document"
  >
    <div
      class="modal-content"
    >
      <div
        class="modal-header"
      >
        <h5
          class="modal-title"
        >
          Find images
        </h5>
        <button
          aria-label="Close"
          class="close"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Ã—
          </span>
        </button>
      </div>
      <div
        class="modal-body"
      >
        <p>
          Browse images from Unsplash.com by searching below.
        </p>
        <div
          class="form-group row mx-0"
        >
          <label
            class="col-form-label mr-2"
            for="img-search-input"
          >
            Search by keyword
          </label>
          <input
            aria-describedby="searchValidationFeedback"
            class="col col-lg-4 form-control  px-2 mr-2"
            id="img-search-input"
            type="text"
            value="cat"
          />
          <button
            class="btn btn-dark mr-auto"
            type="button"
          >
            Search
          </button>
          <div
            class="w-100 d-lg-none"
          />
          <div
            class="col invalid-feedback align-self-center"
            id="searchValidationFeedback"
          >
            Please enter a search term.
          </div>
        </div>
        <div
          class="search-results-container text-center"
        >
          <div
            class="search-results"
          >
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image10-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user10-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User10 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link10.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image11-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user11-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User11 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link11.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result result-selected"
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image12-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user12-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User12 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link12.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image13-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user13-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User13 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link13.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image14-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user14-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User14 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link14.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image15-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user15-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User15 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link15.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image16-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user16-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User16 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link16.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex-sm-nowrap modal-footer"
      >
        <nav
          aria-label="Search results pages"
          class="custom-pagination container mr-sm-3"
        >
          <div
            class="row justify-content-between align-items-center"
          >
            <div
              class="col-auto p-0"
            >
              <button
                class="btn btn-link"
                type="button"
              >
                &lt; Back
              </button>
            </div>
            <div
              class="col-auto"
            >
              <ul>
                <li
                  class="invisible"
                />
                <li>
                  <button
                    class="btn btn-link"
                    type="button"
                  >
                    1
                  </button>
                </li>
                <li>
                  <span
                    class="btn"
                    style="cursor: default;"
                  >
                    2
                  </span>
                </li>
                <li
                  class="invisible"
                />
                <li
                  class="invisible"
                />
              </ul>
            </div>
            <div
              class="col-auto p-0"
            >
              <button
                class="btn btn-link invisible"
                type="button"
              >
                More &gt;
              </button>
            </div>
          </div>
        </nav>
        <button
          class="btn btn-dark flex-shrink-0"
          type="button"
        >
          Use selected image
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`happy path works: snapshot with search results 1`] = `
<div
  class="modal fade"
  role="dialog"
  style="display: block;"
  tabindex="-1"
>
  <div
    class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
    role="document"
  >
    <div
      class="modal-content"
    >
      <div
        class="modal-header"
      >
        <h5
          class="modal-title"
        >
          Find images
        </h5>
        <button
          aria-label="Close"
          class="close"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Ã—
          </span>
        </button>
      </div>
      <div
        class="modal-body"
      >
        <p>
          Browse images from Unsplash.com by searching below.
        </p>
        <div
          class="form-group row mx-0"
        >
          <label
            class="col-form-label mr-2"
            for="img-search-input"
          >
            Search by keyword
          </label>
          <input
            aria-describedby="searchValidationFeedback"
            class="col col-lg-4 form-control  px-2 mr-2"
            id="img-search-input"
            type="text"
            value="cat"
          />
          <button
            class="btn btn-dark mr-auto"
            type="button"
          >
            Search
          </button>
          <div
            class="w-100 d-lg-none"
          />
          <div
            class="col invalid-feedback align-self-center"
            id="searchValidationFeedback"
          >
            Please enter a search term.
          </div>
        </div>
        <div
          class="search-results-container text-center"
        >
          <div
            class="search-results"
          >
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image0-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user0-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User0 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link0.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result result-selected"
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image1-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user1-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User1 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link1.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image2-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user2-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User2 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link2.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image3-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user3-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User3 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link3.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image4-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user4-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User4 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link4.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image5-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user5-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User5 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link5.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image6-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user6-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User6 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link6.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image7-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user7-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User7 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link7.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image8-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user8-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User8 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link8.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="result-container mb-2"
              data-testid="result"
              tabindex="0"
            >
              <div
                class="result "
              >
                <div
                  class="placeholder"
                  style="padding-top: 177.77777777777777%;"
                >
                  <img
                    alt="alt text"
                    src="/image9-small.jpg"
                  />
                  <span
                    class="badge badge-dark"
                  >
                    By 
                    <a
                      class="text-light"
                      href="/user9-link.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      User9 Name
                    </a>
                     
                    on 
                    <a
                      class="text-light"
                      href="/image-link9.html"
                      rel="external noreferrer"
                      target="_blank"
                    >
                      Unsplash
                    </a>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex-sm-nowrap modal-footer"
      >
        <nav
          aria-label="Search results pages"
          class="custom-pagination container mr-sm-3"
        >
          <div
            class="row justify-content-between align-items-center"
          >
            <div
              class="col-auto p-0"
            >
              <button
                class="btn btn-link invisible"
                type="button"
              >
                &lt; Back
              </button>
            </div>
            <div
              class="col-auto"
            >
              <ul>
                <li
                  class="invisible"
                />
                <li
                  class="invisible"
                />
                <li>
                  <span
                    class="btn"
                    style="cursor: default;"
                  >
                    1
                  </span>
                </li>
                <li>
                  <button
                    class="btn btn-link"
                    type="button"
                  >
                    2
                  </button>
                </li>
                <li
                  class="invisible"
                />
              </ul>
            </div>
            <div
              class="col-auto p-0"
            >
              <button
                class="btn btn-link"
                type="button"
              >
                More &gt;
              </button>
            </div>
          </div>
        </nav>
        <button
          class="btn btn-dark flex-shrink-0"
          type="button"
        >
          Use selected image
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with no results 1`] = `
<div
  class="modal fade"
  role="dialog"
  style="display: block;"
  tabindex="-1"
>
  <div
    class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
    role="document"
  >
    <div
      class="modal-content"
    >
      <div
        class="modal-header"
      >
        <h5
          class="modal-title"
        >
          Find images
        </h5>
        <button
          aria-label="Close"
          class="close"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Ã—
          </span>
        </button>
      </div>
      <div
        class="modal-body"
      >
        <p>
          Browse images from Unsplash.com by searching below.
        </p>
        <div
          class="form-group row mx-0"
        >
          <label
            class="col-form-label mr-2"
            for="img-search-input"
          >
            Search by keyword
          </label>
          <input
            aria-describedby="searchValidationFeedback"
            class="col col-lg-4 form-control  px-2 mr-2"
            id="img-search-input"
            type="text"
            value="cat"
          />
          <button
            class="btn btn-dark mr-auto"
            type="button"
          >
            Search
          </button>
          <div
            class="w-100 d-lg-none"
          />
          <div
            class="col invalid-feedback align-self-center"
            id="searchValidationFeedback"
          >
            Please enter a search term.
          </div>
        </div>
        <div
          class="search-results-container text-center"
        >
          <div
            class="text-secondary"
          >
            No results found. :(
          </div>
          <div
            class="search-results"
          />
        </div>
      </div>
      <div
        class="flex-sm-nowrap modal-footer"
      >
        <button
          class="btn btn-dark flex-shrink-0"
          disabled=""
          type="button"
        >
          Use selected image
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders error message correctly (proxy error) 1`] = `
<div
  class="search-error alert alert-warning text-center"
  role="alert"
>
  <h6
    class="alert-heading"
  >
    Something went wrong :0
  </h6>
  <p>
    <small
      class=""
    >
      Error details: 
      Status code 400 (proxy)
       - Bad Request
    </small>
  </p>
  <p>
    Please try again. If you keep getting an error, feel free to 
     
    <a
      class="alert-link"
      href="https://github.com/turtleseason/jigsaw-puzzle/issues"
      rel="external"
    >
      open an issue on GitHub
    </a>
     
    and let me know.
  </p>
  <p
    class="mt-n2"
  >
    You can still search photos directly on 
    <a
      href="https://unsplash.com/"
      rel="external noreferrer"
      target="_blank"
    >
      Unsplash.com
    </a>
    ,
     
    download an image to your computer, and load it using the file selector.
  </p>
</div>
`;

exports[`renders error message correctly (upstream error) 1`] = `
<div
  class="search-error alert alert-warning text-center"
  role="alert"
>
  <h6
    class="alert-heading"
  >
    Something went wrong :0
  </h6>
  <p>
    <small
      class=""
    >
      Error details: 
      Status code 400 (upstream)
       - Pretend something bad happened
    </small>
  </p>
  <p>
    Please try again. If you keep getting an error, feel free to 
     
    <a
      class="alert-link"
      href="https://github.com/turtleseason/jigsaw-puzzle/issues"
      rel="external"
    >
      open an issue on GitHub
    </a>
     
    and let me know.
  </p>
  <p
    class="mt-n2"
  >
    You can still search photos directly on 
    <a
      href="https://unsplash.com/"
      rel="external noreferrer"
      target="_blank"
    >
      Unsplash.com
    </a>
    ,
     
    download an image to your computer, and load it using the file selector.
  </p>
</div>
`;

exports[`renders rate limit error message correctly (proxy) 1`] = `
<div
  class="search-error alert alert-warning text-center"
  role="alert"
>
  <h6
    class="alert-heading"
  >
    Rate limit exceeded (API Proxy)
  </h6>
  <p>
    Please try again in a moment.
    <br />
    (If you get this error frequently, feel free to 
     
    <a
      class="alert-link"
      href="https://github.com/turtleseason/jigsaw-puzzle/issues"
      rel="external"
    >
      open an issue on GitHub
    </a>
     
    and let me know.)
  </p>
  <p
    class="mt-n2"
  >
    You can still search photos directly on 
    <a
      href="https://unsplash.com/"
      rel="external noreferrer"
      target="_blank"
    >
      Unsplash.com
    </a>
    ,
     
    download an image to your computer, and load it using the file selector.
  </p>
</div>
`;

exports[`renders rate limit error message correctly (upstream) 1`] = `
<div
  class="search-error alert alert-warning text-center"
  role="alert"
>
  <h6
    class="alert-heading"
  >
    Rate limit exceeded (API)
  </h6>
  <p>
    The hourly limit for requests from this site to the Unsplash server has been reached.
  </p>
  <p>
    Feel free to 
    <a
      class="alert-link"
      href="https://github.com/turtleseason/jigsaw-puzzle/issues"
      rel="external"
    >
      open an issue on GitHub
    </a>
     
    to let me know - 
    <br />
    if this happens often, I can look into applying for a higher rate limit.
  </p>
  <p
    class="mt-n2"
  >
    You can still search photos directly on 
    <a
      href="https://unsplash.com/"
      rel="external noreferrer"
      target="_blank"
    >
      Unsplash.com
    </a>
    ,
     
    download an image to your computer, and load it using the file selector.
  </p>
</div>
`;
